kani-compiler/src/kani_middle/transform/loop_contracts.rs

```diff
@@ -0,0 +1,467 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! This module contains code related to the MIR-to-MIR pass to enable loop contracts.
+//!
+
+use crate::kani_middle::KaniAttributes;
+use crate::kani_middle::find_fn_def;
+use crate::kani_middle::transform::body::{InsertPosition, MutableBody, SourceInstruction};
+use crate::kani_middle::transform::{BodyTransformation, CallGraph, TransformationResult};
+use crate::kani_queries::QueryDb;
+use crate::stable_mir::CrateDef;
+use rustc_middle::ty::TyCtxt;
+use rustc_span::Symbol;
+use stable_mir::DefId;
+use stable_mir::mir::mono::{Instance, MonoItem};
+use stable_mir::mir::{
+    BasicBlock, BasicBlockIdx, Body, ConstOperand, Operand, Rvalue, Statement, StatementKind,
+    Terminator, TerminatorKind,
+};
+use stable_mir::ty::{FnDef, MirConst, RigidTy};
+use std::collections::VecDeque;
+use std::collections::{HashMap, HashSet};
+use std::fmt::Debug;
+
+use super::GlobalPass;
+
+/// This pass will perform the following operations:
+/// 1. Replace the body of `kani_register_loop_contract` by `kani::internal::run_contract_fn`
+///    to invoke the closure.
+///
+/// 2. Transform loops with contracts from
+///    ```ignore
+///    bb_idx: {
+///         loop_head_stmts
+///         _v = kani_register_loop_contract(move args) -> [return: terminator_target];
+///    }
+///
+///    ...
+///    loop_body_blocks
+///    ...
+///
+///    loop_latch_block: {
+///         loop_latch_stmts
+///         goto -> bb_idx;
+///    }
+///    ```
+///    to blocks
+///    ```ignore
+///    bb_idx: {
+///         _v = true
+///         goto -> terminator_target
+///    }
+///
+///    ...
+///    loop_body_blocks
+///    ...
+///
+///    loop_latch_block: {
+///         loop_latch_stmts
+///         goto -> bb_new_loop_latch;
+///    }
+///
+///    bb_new_loop_latch: {
+///         loop_head_body
+///         _v = kani_register_loop_contract(move args) -> [return: bb_idx];
+///    }
+///    ```
+///
+/// 3. Move the statements `loop_head_stmts` of loop head  with contracts to the body of register
+///    functions for later codegen them as statement_expression in CBMC loop contracts.
```

Comment:
Sorry, how does this work? What about the local declarations? How do we know this is sound?